### ISSUE TITLE
Why gradient does not scale linearly with CTCloss?

### ISSUE THREAD

**Issue by lifelongeek** [SeanNaren/deepspeech.pytorch#250]
I am trying to use CTCLoss with another loss, and want coefficients of CTCLoss smaller.
But when I tried different 

This is the extreme cases for small CTCLoss. (CTCLoss coefficient = 0)

```
criterion = CTCLoss()
loss = 0*criterion(out, targets, sizes, target_sizes)
model.zero_grad()
loss.backward()
optimizer.step()

for param in model.parameters():
     if param.grad is not None:
          grad_norm += param.grad.norm().data[0]
```

In above example, loss is always 0. 
However, fairly large gradient value still exists even CTCLoss coefficient is 0.

Can anyone explain why gradient does not scale linearly with CTCLoss?

**Comment by SeanNaren** [comment]
I can point people in the direction, however I'm unsure of the fix.

If you look [here](https://github.com/SeanNaren/warp-ctc/blob/pytorch_bindings/pytorch_binding/warpctc_pytorch/__init__.py#L50) we don't take grad_outputs into consideration when calculating the gradients, thus any scale factor isn't being applied to the gradients.

There was a commit added to the repo to fix this but I reverted it (see commit [here](https://github.com/SeanNaren/warp-ctc/commit/d6f192cb256ba403b8e27bccf5a22e67fefe78a8). The reason I reverted is because convergence became significantly longer when this fix was added.

Any help on this would be great :)

cc @weedwind

**Comment by ssorcerer** [comment]
Does this solve the problem?
    @staticmethod
    def backward(ctx, grad_output):
        print(ctx.grads, grad_output)
        grad_output = grad_output.to(ctx.grads.device)
        return ctx.grads * grad_output, None, None, None

**Comment by SeanNaren** [comment]
@jinserk sorry for being away, was lurking on the forums and stuff!

We've been running into the same issue internally; the gradients not being scaled has caused an issue for us.

The issue right now is that convergence seems to be better when the gradients are not scaled, however this is less a feature but more a bug. However when we fix this it causes [this](https://github.com/SeanNaren/deepspeech.pytorch/blob/master/train.py#L252) line to scale the gradient to the batch size. This seems to reduce the speed of convergence dramatically.

The proper thing to do is scale the gradients by the batch size. This will have to alter the learning rates to match this, and this is something I haven't looked at so far! The paddle paddle DeepSpeech repo swapped to ADAM which might help with since no learning rate is really needed to be specified.

**Comment by jinserk** [comment]
Thank you for replying, @SeanNaren! When I used the official `nn.CTCLoss`, the loss itself reduced so fast, but the accuracy stayed not being trained at all. I tried the Adam too, but the result was almost same. Loss was reduced quicker, but the accuracy didn't get better. 
If I give the arg `size_average=True`, is it already scaling the gradient by the batch size?

**Comment by SeanNaren** [comment]
@jinserk oh god, any idea if the softmax happens internally in the torch version? This would mess things up a bit!

**Comment by jinserk** [comment]
`nn.CTCLoss` doesn't have softmax in it. We need to give explicit log softmax of the input to it. I've tested but this isn't making the difference. I guess the most difference was happened by the gradient connection.

**Comment by SeanNaren** [comment]
There has been a few advancements in the PyTorch CTCLoss function, however the underlying issue is still the same.

Since the gradients are not properly scaled it seems like when warp-ctc was used by Baidu they've set up their hyper-parameters (and this library) to match this bug. When scaling appropriately occurs, this causes a change required in the hyper-parameters.

This has become obvious when we scale or need to use dynamic loss scaling from [apex](https://nvidia.github.io/apex/fp16_utils.html?highlight=dynamic#apex.fp16_utils.DynamicLossScaler) since our gradients become very large and incorrectly averaged.

To fix this hyper-parameters will need to be re-tuned once length averaging is used and the gradient bug is fixed.

**Comment by ppriyank** [comment]
So what was the final conclusion for **nn.CTCLoss()** ??? 
Can I just do 

```
out, output_sizes = model(temp, input_sizes)
out = out.transpose(0, 1)
criterion = nn.CTCLoss()
loss = criterion(float_out, targets, output_sizes, target_sizes).to(device)
```

and eventually `scaled_loss.backward()`
or do I need to something else as well, like scaling gradients, softmax, etc??

**Comment by ppriyank** [comment]
I keep getting the "nan" error (nan gradient) in the model when using `nn.CTCLoss()`
I'm planning to modify the original audio as well so used torch stft

```
window_size = audio_conf['window_size']
sample_rate = audio_conf['sample_rate']
window_stride = audio_conf['window_stride']
n_fft = int(sample_rate * window_size)
win_length = n_fft
hop_length = int(sample_rate * window_stride)
stft = STFT(filter_length=n_fft,  hop_length=hop_length, win_length=win_length,window='hamming').to(device)
```

```
audio  = load_audio(path)
original = torch.FloatTensor(audio).to(device)
original.requires_grad= False
model.requires_grad = True
model.zero_grad()

original = original.unsqueeze(0)
magnitude, phase = stft.transform(original) ## torch.Size([1, 161, 989])
magnitude = magnitude.unsqueeze(0)
magnitude = magnitude.to(device)
temp = torch.log(magnitude + 1)
input_sizes = torch.IntTensor([temp.size(3)]).int() ## 989
out, output_sizes = model(temp, input_sizes) 
## out:  torch.Size([1, 495, 29])   , output_sizes: 495


decoder = GreedyDecoder(model.labels, blank_index=model.labels.index('_'))
decoded_output, decoded_offsets = decoder.decode(out, output_sizes)
print(decoded_output)
# [['AN HENCE WE FIND THE SAME SORT OF CLUMSINESS IN THE TIMEUS OF PLATO WHICH CHARACTERIZES THE PHILOSOPHICAL PORM OF LECRETIOUS']]
```

transcript is : 
`'AND HENCE WE FIND THE SAME SORT OF CLUMSINESS IN THE TIMAEUS OF PLATO WHICH CHARACTERIZES THE PHILOSOPHICAL POEM OF LUCRETIUS'`

```
criterion = nn.CTCLoss().to(device)
out = out.transpose(0, 1)  # TxNxH
float_out = out.float()  # ensure float32 for loss
targets = torch.tensor(int_transcript , dtype= torch.int32 )
target_sizes = torch.tensor([len(transcript)] , dtype= torch.int32)
loss = criterion(softmax(float_out), targets, output_sizes, target_sizes)
optimizer_sound.zero_grad()
loss.backward()
for param in model.parameters():
	param.grad
	break 
```

all gradients are `nan`, not sure what am I doing wrong....
